lib/rucio/client/downloadclient.py

```diff
@@ -73,6 +73,134 @@ def __init__(self, _client=None, user_agent='rucio_clients', logger=None):
+            ignore_checksum     - Optional: If true, the checksum validation is skipped (for pfn downloads the checksum must be given explicitly). (Default: True)
+            transfer_timeout    - Optional: Timeout time for the download protocols. (Default: None)
+        :param num_threads: Suggestion of number of threads to use for the download. It will be lowered if it's too high.
+        :param trace_custom_fields: Custom key value pairs to send with the traces
+
+        :returns: a list of dictionaries with an entry for each file, containing the input options, the did, and the clientState
+                  clientState can be one of the following: ALREADY_DONE, DONE, FILE_NOT_FOUND, FAIL_VALIDATE, FAILED
+
+        :raises InputValidationError: if one of the input items is in the wrong format
+        :raises NoFilesDownloaded: if no files could be downloaded
+        :raises NotAllFilesDownloaded: if not all files could be downloaded
+        :raises RucioException: if something unexpected went wrong during the download
+        """
+        logger = self.logger
+        trace = copy.deepcopy(self.trace_tpl)
+        log_prefix = 'Unarchiving'
+
+        logger.info('Processing %d item(s) for input' % len(items))
+        input_items = []
+        for item in items:
+            archive = item.get('archive')
+            file_extract = item.get('did')
+            rse_name = item.get('rse')
+            file_extract_scope, file_extract_name = self._split_did_str(file_extract)
+            archive_scope, archive_name = self._split_did_str(archive)
+
+            # if not specified, taking first availbale replica
+            if not rse_name:
+                replicas = self.client.list_replicas([{'scope': archive_scope, 'name': archive_name}],
+                                                     schemes='root',
+                                                     client_location=detect_client_location())
+                for r in replicas:
+                    for rse in r['states'].keys():
+                        if r['states'][rse] == 'AVAILABLE':
+                            rse_attr = self.client.list_rse_attributes(rse)
+                            if rse_attr['istape'] == 'False':
+                                rse_name = rse
+                                break
+                    if rse_name and rse_name is not None:
+                        break
+
+            rse = {}
+            try:
+                rse = rsemgr.get_rse_info(rse_name)
+            except RSENotFound:
+                logger.warning('%sCould not get info of RSE %s' % (log_prefix, rse_name))
+                continue
+
+                raise InputValidationError('The keys archive (zip file), file_extract, and rse are mandatory')
+
+            logger.debug('Preparing pfn of archive for download of %s (%s) from %s' % (archive, file_extract, rse))
+            pfn_archive = ''
+            try:
+                pfns_dict = rsemgr.lfns2pfns(rse,
```

Comment:
from some reason pfns is not populated, and also I would need to do mapping between pfn and rse:

output of list_file_replicas:

{u'adler32': u'48e8ce74', u'name': u'DAOD_BPHY5.13781702._000102.zip.1', u'rses': {u'CERN-PROD_DATADISK': [], u'BNL-OSG2_DATADISK': [], u'UNI-FREIBURG_DATADISK': [], u'PRAGUELCG2_DATADISK': []}, u'bytes': 1106077679, u'states': {u'CERN-PROD_DATADISK': u'AVAILABLE', u'BNL-OSG2_DATADISK': u'AVAILABLE', u'UNI-FREIBURG_DATADISK': u'AVAILABLE', u'PRAGUELCG2_DATADISK': u'AVAILABLE'}, u'pfns': {}, u'scope': u'data16_valid', u'md5': None}

Reply:
Don't know why PFNs are not populated but I don't think you would need this mapping.
Either you get the RSE from the input options or you try to use RSE after RSE. List replicas should already give them in a sorted order.

Reply:
ok, but for now I leave rather as it is, since it is working. Once we figure out, why pfns are not populated, we can remove the lfns2pfns part from the code.

Reply:
if pfns are empty, then the protocol does not exist for the RSE.

and i now see the bug:      `schemes='root'` should be  `schemes=['root']`